### PR TITLE
http-client: Deserialize Map<K, V> with Jackson properly

### LINKED ISSUE #6011: http-client cannot deserialize Map<K, V>
### Expected Behavior

Deserialize `Map<K, V>` properly with declarative http client.

https://github.com/micronaut-projects/micronaut-core/blob/f5e363587b4bb46dd90904e86028c5ef97c0798e/runtime/src/main/java/io/micronaut/jackson/codec/JacksonMediaTypeCodec.java#L175
Here raw type is given to jackson reader, however json stream can properly read by byte[] and `JavaType` such as:

```
JavaType javaType = constructJavaType(type);
getObjectMapper().readValue(buffer.toByteArray(), javaType)
```

since we have `Argument<T> type` available.

### Actual Behaviour

```
ERROR i.m.h.s.netty.RoutingInBoundHandler - Unexpected error occurred: class java.util.LinkedHashMap cannot be cast to class com.migros.generatedapi.model.PricingResultDto (java.util.LinkedHashMap is in module java.base of loader 'bootstrap'; com.migros.generatedapi.model.PricingResultDto is in unnamed module of loader 'app')
java.lang.ClassCastException: class java.util.LinkedHashMap cannot be cast to class com.migros.generatedapi.model.PricingResultDto (java.util.LinkedHashMap is in module java.base of loader 'bootstrap'; com.migros.generatedapi.model.PricingResultDto is in unnamed module of loader 'app')
	at com.migros.TestController.testGet(TestController.java:32)
	at com.migros.$TestControllerDefinition$$exec1.invokeInternal(Unknown Source)
	at io.micronaut.context.AbstractExecutableMethod.invoke(AbstractExecutableMethod.java:151)
	at io.micronaut.context.DefaultBeanContext$4.invoke(DefaultBeanContext.java:482)
	at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:246)
	at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:121)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.emitRouteResponse(RoutingInBoundHandler.java:1502)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.access$2100(RoutingInBoundHandler.java:151)
	at io.micronaut.http.server.netty.RoutingInBoundHandler$6$1.request(RoutingInBoundHandler.java:1469)
	at io.micronaut.http.server.netty.async.ContextCompletionAwareSubscriber.doOnSubscribe(ContextCompletionAwareSubscriber.java:49)
	at io.micronaut.core.async.subscriber.CompletionAwareSubscriber.onSubscribe(CompletionAwareSubscriber.java:38)
	at io.micronaut.http.server.netty.RoutingInBoundHandler$6.doSubscribe(RoutingInBoundHandler.java:1457)
	at io.micronaut.http.server.netty.RoutingInBoundHandler$6.subscribe(RoutingInBoundHandler.java:1448)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.executeRoute(RoutingInBoundHandler.java:1119)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.handleRouteMatch(RoutingInBoundHandler.java:775)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.channelRead0(RoutingInBoundHandler.java:606)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.channelRead0(RoutingInBoundHandler.java:148)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:102)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.micronaut.http.netty.stream.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:200)
	at io.micronaut.http.netty.stream.HttpStreamsServerHandler.channelRead(HttpStreamsServerHandler.java:123)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.flow.FlowControlHandler.dequeue(FlowControlHandler.java:200)
	at io.netty.handler.flow.FlowControlHandler.channelRead(FlowControlHandler.java:162)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:831)

```

https://github.com/micronaut-projects/micronaut-core/blob/f5e363587b4bb46dd90904e86028c5ef97c0798e/runtime/src/main/java/io/micronaut/jackson/codec/JacksonMediaTypeCodec.java#L175 here we have full generic type information but not using.

### Steps To Reproduce

Lets say we have a controller handler:
```
@Get("/pricing/{productId}")
  public Mono<Map<String, PricingResultDto>> getSearchResponse(@PathVariable Long productId) {
    return Mono.just(Map.of("test1", new PricingResultDto().price(new AmountDto().currency(CurrencyCode.EUR)),
                            "test2", new PricingResultDto().price(new AmountDto().currency(CurrencyCode.USD))));
  }
```
and generated client accordingly like:
```

@Client(id="test-generated")
public interface DefaultApi {
 
  @Get("/pricing/{productId}")
  Flux<Map<String, PricingResultDto>> getProductPricing(
    @PathVariable("productId") Long productId);
  }
```

When we inject and call `defaultApi.getProductPricing(123L).blockLast()`, server error above occurs.

Whole test codebase is available via github link (generated declarative client is in `com.migros.generatedapi` package). It can be easily tested with `curl -v localhost:8080/test-api`.

### Environment Information

- OSX 10.15.7
- openjdk 16.0.1 2021-04-20

### Example Application

https://github.com/nyilmaz/mcnt-demo

### Version

2.5.11

### PR DESCRIPTION
**Pull request by nyilmaz**
Addresses https://github.com/micronaut-projects/micronaut-core/issues/6011

### CODE REVIEW

File: runtime/src/main/java/io/micronaut/jackson/codec/JacksonMediaTypeCodec.java [hunk 1]
```diff
@@ -172,7 +173,9 @@ public boolean supportsType(Class<?> type) {
      */
     public <T> T decode(Argument<T> type, JsonNode node) throws CodecException {
         try {
-            return getObjectMapper().treeToValue(node, type.getType());
+            ObjectMapper objectMapper = getObjectMapper();
+            JsonParser jsonParser = objectMapper.treeAsTokens(node);
+            return objectMapper.readValue(jsonParser, constructJavaType(type));
```
Comment (jameskleeh) [inline runtime/src/main/java/io/micronaut/jackson/codec/JacksonMediaTypeCodec.java]: Please wrap this in a `if (type.hasTypeVariables()) {` as is done above to limit the potential for different behavior

Comment (nyilmaz) [reply runtime/src/main/java/io/micronaut/jackson/codec/JacksonMediaTypeCodec.java]: sorry, missed your review, done

**Comment by jameskleeh** [comment]
Thanks!

**Comment by cemo** [comment]
@jameskleeh can we cut a release please?

**Pull request merged**